### ISSUE TITLE
7.1.x breaks TLS behavior

### ISSUE THREAD

**Issue by calavera** [apache/trafficserver#2129]
I'm still seen inconsistencies behavior between 7.0 and 7.1 negotiating TLS certificates.

When I try to resolve a domain that doesn't exist, this is the client log with 7.0:

```
`curl  -s -vv --resolve "www.nothing.com:443:127.0.0.1" https://www.nothing.com > /dev/null`
* Added www.nothing.com:443:127.0.0.1 to DNS cache
* Rebuilt URL to: https://www.nothing.com/
* Hostname was found in DNS cache
*   Trying 127.0.0.1...
* Connected to www.nothing.com (127.0.0.1) port 443 (#0)
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* SSLv3, TLS handshake, Client hello (1):
} [data not shown]
* SSLv3, TLS alert, Server hello (2):
{ [data not shown]
* error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure
* Closing connection 0
```

This is the client log with 7.1:

```
`curl  -s -vv --resolve "www.nothing.com:443:127.0.0.1" https://www.nothing.com > /dev/null`
* Added www.nothing.com:443:127.0.0.1 to DNS cache
* Rebuilt URL to: https://www.nothing.com/
* Hostname was found in DNS cache
*   Trying 127.0.0.1...
* Connected to www.nothing.com (127.0.0.1) port 443 (#0)
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* SSLv3, TLS handshake, Client hello (1):
} [data not shown]
* Unknown SSL protocol error in connection to www.nothing.com:443 
* Closing connection 0
```

As you can see the final error is different.

I have a plugin with a hook for `TS_CERT_NAME`. This plugin lazy loads certificates from an external API. It uses HttpAsync to perform the req/resp and it re-enables the continuation after the response arrives from the server.

Everything works as expected in 7.0, however, in 7.1 the first certificate resolution always fails with the same error I posted above, but the next resolution works. My hunch is that it works because the certificate in the second request is cached in memory, we use an LRU structure, but the first request doesn't respect the re-enabling flow.

This is the first client request:

```
curl  -s -vv --resolve "www.netlify.com:443:127.0.0.1" https://www.netlify.com           
* Added www.netlify.com:443:127.0.0.1 to DNS cache
* Rebuilt URL to: https://www.netlify.com/
* Hostname was found in DNS cache
*   Trying 127.0.0.1...
* Connected to www.netlify.com (127.0.0.1) port 443 (#0)
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* SSLv3, TLS handshake, Client hello (1):
* Unknown SSL protocol error in connection to www.netlify.com:443 
* Closing connection 0
```

This is the second client request:
```
curl  -s -vv --resolve "www.netlify.com:443:127.0.0.1" https://www.netlify.com
* Added www.netlify.com:443:127.0.0.1 to DNS cache
* Rebuilt URL to: https://www.netlify.com/
* Hostname was found in DNS cache
*   Trying 127.0.0.1...
* Connected to www.netlify.com (127.0.0.1) port 443 (#0)
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* SSLv3, TLS handshake, Client hello (1):
* SSLv3, TLS handshake, Server hello (2):
* SSLv3, TLS handshake, CERT (11):
* SSLv3, TLS handshake, Server key exchange (12):
* SSLv3, TLS handshake, Server finished (14):
* SSLv3, TLS handshake, Client key exchange (16):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv3, TLS handshake, Finished (20):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv3, TLS handshake, Finished (20):
* SSL connection using ECDHE-RSA-AES128-GCM-SHA256
* Server certificate:
* 	 subject: OU=Domain Control Validated; CN=*.netlify.com
* 	 start date: 2016-11-23 18:28:10 GMT
* 	 expire date: 2018-11-24 18:28:10 GMT
* 	 subjectAltName: www.netlify.com matched
* 	 issuer: C=BE; O=GlobalSign nv-sa; CN=AlphaSSL CA - SHA256 - G2
* 	 SSL certificate verify ok.
> GET / HTTP/1.1
> User-Agent: curl/7.35.0
> Host: www.netlify.com
> Accept: */*
> 
< HTTP/1.1 200 OK
< Cache-Control: public, max-age=0, must-revalidate
< Content-Type: text/html; charset=UTF-8
< Date: Thu, 08 Jun 2017 23:06:11 GMT
< Etag: "2514ae6fa3d05bd77ad134091908ba0e-ssl"
< Strict-Transport-Security: max-age=31536000
< Age: 425262
< Content-Length: 38495
< Connection: keep-alive
* Server Netlify is not blacklisted
< Server: Netlify
< 
```

I uploaded the full log to this pastebin: https://pastebin.com/dA5GTHGp
I just built from the 7.1.x branch with this SHA: https://github.com/apache/trafficserver/commit/7f1c3943222f811a79dcf19021562826bb6986f1
My OpenSSL version is 1.0.2k

**Comment by masaori335** [comment]
Do you see something like below in debug log with 'ssl.*' tag? If so, #2086 might fix.

```
[Jun 14 08:58:08.341] Server {0xb0016000} DEBUG: <SSLNetVConnection.cc:1067 (sslServerHandShakeEvent)> (ssl) 0x981bb70 first read
[Jun 14 08:58:08.341] Server {0xb0016000} DEBUG: <SSLNetVConnection.cc:374 (read_raw_data)> (ssl) 0x981bb70 read r=0 total=4096 bio=0
[Jun 14 08:58:08.341] Server {0xb0016000} DEBUG: <SSLNetVConnection.cc:1083 (sslServerHandShakeEvent)> (ssl-diag) SSL handshake error: EOF
```

**Comment by calavera** [comment]
@masaori335 yes, I see those lines at the end of the log in https://pastebin.com/dA5GTHGp. However, it doesn't look like you patch helped 😔 

```
[Jun 13 21:34:34.439] Server {0x2b4a1328d700} DEBUG: <SSLNetVConnection.cc:1067 (sslServerHandShakeEvent)> (ssl) 0x2b4a5401bca0 first read
[Jun 13 21:34:34.439] Server {0x2b4a1328d700} DEBUG: <SSLNetVConnection.cc:374 (read_raw_data)> (ssl) 0x2b4a5401bca0 read r=0 total=3799 bio=0
[Jun 13 21:34:34.439] Server {0x2b4a1328d700} DEBUG: <SSLNetVConnection.cc:1083 (sslServerHandShakeEvent)> (ssl-diag) SSL handshake error: EOF
```

**Comment by calavera** [comment]
Tracking @masaori335's comment here from #2086 regarding this new log https://pastebin.com/Kqp9n2e3

> read_raw_data() is called twice and first one could read 297 bytes. And second one has nothing to read (it's not problem). Probably all of TLS Client Hello is loaded in the handShakeBuffer by first one.
IMO, the problem is that the TLS Client Hello is not used after VConn is reenabled by your plugin.

**Comment by calavera** [comment]
After much trial an error, this is what it seems to fix the problem, matching the behavior I see in 7.0.x:

1. Revert https://github.com/apache/trafficserver/commit/0e5090bdea5b988f4f97b1d95801e92e6b0e13f9 from the 7.1.x branch
2. Apply #2152 after 1

/cc @shinrich, @zwoop

**Comment by zwoop** [comment]
Do we still need the revert of 0e5090b, or is your fix in #2152 sufficient? If not, we obviously need another fix here for 0e5090b as well.

**Comment by calavera** [comment]
> is your fix in #2152 sufficient?

that fix is not enough. Sorry, I didn't arrive to understand the changes in `0e5090b` and I didn't have a better solution than reverting it. They are two separate problems.

The current HEAD in 7.1.x and master don't work with my plugin because of the problem described here. Reverting `0e5090b` fixes the problem but it doesn't surfaces another problem that was present before. #2152 solves the second problem.

**Comment by zwoop** [comment]
Do we have any ideas what additional fix(es) are needed? @shinrich ?

**Comment by calavera** [comment]
As a note here, the problems described in this issue are fixed by #2209.

**Comment by zwoop** [comment]
I'm going to close this, unless you have issues other than #2209 (which will go into 7.1.1).
